Initial version of jest-worker

mjesun opened the pull request:
This PR introduces a new module, `jest-worker`, intended to allow heavy task parallelization over multiple workers. 

The module has a few advantages over the currently one used both in `jest` and `metro-bundler`:

* 100% `flow`-ified.

* 100% test coverage on it, all statements, methods and branches.

* Slightly faster than the currenly used one.

* Natively provides a `Promise` based interface, which allow us to avoid the extra wrapping layer in order to be used with `async`/`await.

* It only has one single dependency (`merge-stream`), which we could also remove.

* Lazily instantiated code in worker, meaning no code is loaded on child processes until the first call is done (lazy require). This allows to spawn a farm with minimal RAM consumption, and only load them when needed.

* Sticky workers: tasks will be processed by the first available worker if no stickyness is needed, or by a particular one if the task is forced to do so. Specially useful for workers implementing caches.


## Performance test

It can be run by doing `node --expose-gc test.js` under `__performance_tests__`. Note that the percentage improvement shown (~ 10%) applies to 10,000 calls, meaning the performance improvement per single call is negligible. The test implements a `Promise` wrapper over the current implementation, so we can equivalently test both implementations as we use them in real scenarios.

```
---------------------------------------------------------------------------
jest-worker: { globalTime: 738, processingTime: 707 }
worker-farm: { globalTime: 885, processingTime: 866 }
---------------------------------------------------------------------------
jest-worker: { globalTime: 738, processingTime: 718 }
worker-farm: { globalTime: 865, processingTime: 849 }
---------------------------------------------------------------------------
jest-worker: { globalTime: 708, processingTime: 685 }
worker-farm: { globalTime: 769, processingTime: 753 }
---------------------------------------------------------------------------
jest-worker: { globalTime: 682, processingTime: 656 }
worker-farm: { globalTime: 780, processingTime: 764 }
---------------------------------------------------------------------------
jest-worker: { globalTime: 704, processingTime: 684 }
worker-farm: { globalTime: 775, processingTime: 757 }
---------------------------------------------------------------------------
jest-worker: { globalTime: 705, processingTime: 677 }
worker-farm: { globalTime: 767, processingTime: 748 }
---------------------------------------------------------------------------
jest-worker: { globalTime: 700, processingTime: 675 }
worker-farm: { globalTime: 766, processingTime: 751 }
---------------------------------------------------------------------------
jest-worker: { globalTime: 728, processingTime: 702 }
worker-farm: { globalTime: 770, processingTime: 755 }
---------------------------------------------------------------------------
jest-worker: { globalTime: 721, processingTime: 695 }
worker-farm: { globalTime: 769, processingTime: 756 }
---------------------------------------------------------------------------
jest-worker: { globalTime: 702, processingTime: 675 }
worker-farm: { globalTime: 801, processingTime: 784 }
---------------------------------------------------------------------------
total worker-farm: { wFGT: 7947, wFPT: 7783 }
total jest-worker: { jWGT: 7126, jWPT: 6874 }
---------------------------------------------------------------------------
% improvement over 10000 calls (global time): 10.330942494022901
% improvement over 10000 calls (processing time): 11.679301040729795
```


## Coverage

![Coverage](https://user-images.githubusercontent.com/829269/31123180-badf2e92-a836-11e7-8460-1b5c88719517.png)

mjesun commented:
@aaronabramov Tests failed on Node 4 because `...` is not parseable; any idea why?

thymikee commented:
@mjesun We transform rest/spread only in function declaration params (`transform-es2015-parameters` plugin). You'll need to use `call`/`apply` directly instead.

rafeca commented on packages/jest-parallel/src/index.js:
```diff
@@ -0,0 +1,185 @@
+import type {Readable} from 'stream';
+
+/**
+ * The Metro Bundler farm is a class that allows you to queue methods across
```
s/Metro Bundler farm/JestFarm

mjesun replied:
👍

rafeca commented on packages/jest-parallel/src/worker.js:
```diff
@@ -0,0 +1,166 @@
+    });
+
+    this._child = child;
+    this._busy = false;
```
Does this needed to be here? AFAICT this could go into the constructor directly (if creating a process was asynchronous then it would make sense, but then you would also have to set `busy = true` at the beginning of the initialization and then call `_process()` at the end).

mjesun replied:
The `initialize` method can be called multiple times; not only when instantiating `JestWorker`, but also if the child process unexpectedly dies. Once the process is restarted, it cannot be busy (whatever was the state of the previous process), so that's why it's set here.

rafeca commented on packages/jest-parallel/src/index.js:
```diff
@@ -0,0 +1,185 @@
+    };
+
+    for (let i = 0; i < options.workers; i++) {
+      const metroWorker = new JestWorker(workerOptions);
```
metro... 😄

mjesun replied:
👍

rafeca commented:
Awesome job! I love how easy is to understand the internal logic!

One question: could a user of this farm have sticky workers without having to decide explicitly in which worker gets initially assigned to each call?

Let me explain: As a user, I would prefer not to force `JestFarm` to process each task on a specific worker the first time the task is executed. Instead, I would like `JestFarm` to assign a random worker based on load on the first execution (`JestFarm` will be more optimal by assigning the task to the worker that its free) and stick to it on the next calls (so the cache is reused).

AFAICT this would not be possible with the current API: `JestFarm` would need to expose the worker that has processed each task inside the response, this way the user could use this information in his implementation of `getWorker()`.

This use case would work really well for Metro Bundler, since in general hundreds/thousands of different tasks are processed concurrently on the first load (when building the initial bundle), but only a small amount of concurrent tasks are processed afterwards (incremental builds), which we want to be sticky.

mjesun commented:
@rafeca What you described is exactly how `jest-parallel` works: you can return `null` or any string from the `getWorker` method. If you return `null`, or if you return any string that was never seen before, the task will be executed ASAP by the first available worker, no matter which one it is.

Once executed, the string will be tied to the worker that executed it, so the **next time** you call the task and you return that same string, the task will be executed by the worker that processed your first call.

This allows for optimal parallelization of tasks when there is no worker preference, but it also allows to get a sticked worker when needed. I changed the `README.md` to add some information about how sticky workers work, in order to clarify it.

SimenB commented on packages/jest-parallel/README.md:
```diff
@@ -0,0 +1,107 @@
+The only exposed method is a constructor (`JestFarm`) that is initialized by passing an options object. This object can have the following properties:
+
+
+### `exposedMethods: $ReadOnlyArray<string>` (required)
```
What if you want to call a `module.exports = function() {}`? This is done internally in jest now

I came here to ask about using the babel interop to make transpiled `export default` work, but if a default export is not supported, I guess we can just do `exposedMethods: ['default']`

mjesun replied:
Yep, the way `default` exports are exported by Babel is by adding a `default` key to the object, so it is technically possible to call them with the current implementation.

That said, standard exports can't be called, so this is a fair point. I will change the default exported object of `getApi` from an `Object.create(null)` to a bound `_makeCall` that calls with `default`.

SimenB replied:
Example of where jest needs the default export (can be changed of course):

https://github.com/facebook/jest/blob/95278b2047051244f6cce416c1fbdebcd807265e/packages/jest-haste-map/src/worker.js#L43-L44

mjesun replied:
I modified the code, so `api` is now also a wrapped call :)

SimenB replied:
Awesome!

Any chance of using the babel interop instead of plain `require`, or do you think that'd be confusing behavior?

SimenB replied:
And to clarify, is the way you invoke the main function by passing `exposedMethods: ['']` or `exposedMethods: []`?

mjesun replied:
Yes, I'm fine implementing the Babel logic to be able to treat Babel-ified modules as if they were a standard one. Regarding the `['']`, it is not necessary to provide it, the default export is always exposed (i.e. the result of `getApi()` is always a method). This makes me think that `exposedMethods` should then become optional (there is no sense in making `[]` a required option).

I will also adjust `README.md` to add more information about default exports and potentially provide an example on how to use it.

rafeca commented on packages/jest-parallel/README.md:
```diff
@@ -0,0 +1,107 @@
+
+### `getWorker: (method: string, ...args: Array<any>) => ?string` (optional)
+
+Every time a method exposed via the API is called, `getWorker` is also called in order to stick the call into a worker. This is useful for workers that are able to cache the result or part of it. You stick calls to worker by making `getWorker` return the same identifier for all of them. If you do not want to stick the call to any worker, return `null`.
```
Nice!!! could we have a simple example of usage with sticky workers in the usage section?

rafeca commented:
> @rafeca What you described is exactly how jest-parallel works: you can return null or any string from the getWorker method. If you return null, or if you return any string that was never seen before, the task will be executed ASAP by the first available worker, no matter which one it is.

Ohhh nice!! 😃  I misread the Readme, thanks for making it more clear. Apart from that, I would suggest changing the `getWorker()` method name: I find it misleading since it's not actually returning a worker, but an identifier for each call that will be used to stick calls to specific workers. A suggestion for the new name could be `getStickyId()`, but if you have a better name feel free to use it 😀

SimenB commented on packages/jest-parallel/src/__tests__/worker.test.js:
```diff
@@ -0,0 +1,274 @@
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @format
```
I don't think these are needed in this code base

mjesun replied:
😅 true! These ones come from a copy of an internal codebase. I will remove them 🙂

mjesun commented on packages/jest-parallel/README.md:
```diff
@@ -0,0 +1,107 @@
+Allow customizing all options passed to `childProcess.fork`. By default, some values are set (`cwd` and `env`), but you can override them and customize the rest. For a list of valid values, check [the Node documentation](https://nodejs.org/api/child_process.html#child_process_child_process_fork_modulepath_args_options).
+
+
+### `getWorker: (method: string, ...args: Array<any>) => ?string` (optional)
```
Yes, from all given names the one I like the most is `getWorkerKey`. It is a getter, and tells stuff about the workers. I will adjust all references to have that new name.

mjesun commented:
@SimenB Updated `child.js` so that it is able to use Babel `__esModule`'s now.
@davidaurelio Updated `getWorker` to `getWorkerKey`, as well as the `README.md` file.
@rafeca Provided example on how to use the `getWorkerKey` functionality.

mjesun commented:
The test for Node 4 broke with a `Segmentation fault` :|  Is there any way to re-trigger them?

aaronabramov commented:
there's a "rebuild" button in the top right corner.
i just restarted the build

cpojer commented on packages/jest-parallel/README.md:
```diff
@@ -0,0 +1,173 @@
+# jest-parallel
+
+Module for executing heavy tasks under forked processes in parallel, by providing a modern interface (`Promise` based), minimum overhead, and sticky workers.
```
can we just say "a `Promise` based interface" rather than modern? The word modern only works at one point in time, and will sound dated in a year. Make it timeless.

cpojer commented on packages/jest-parallel/README.md:
```diff
@@ -0,0 +1,173 @@
+
+The module works by providing an absolute path of the module to be loaded in all forked processes, as well as the list of methods that can be remotely called. All methods are exposed on the parent process as promises, so they can be `await`'ed. Child (worker) methods can either be synchronous or asynchronous.
+
+The way sticky workers work is by using the returned string of the `getWorkerKey` method. If the string was used before, the call will be queued to the related worker; if not, it will be executed by the first available worker, then sticked to the worker that executed it; so the next time it will be processed by the same worker. If you have no preference on the worker executing the task, but you h ...
```
Can you explain at first what a sticky worker is meant to be before you explain how to use it? Instead of saying "sticked", can you say "bound"? "it will be executed by the first available worker, then bound to the that same worker for subsequent calls"

cpojer commented on packages/jest-parallel/README.md:
````diff
@@ -0,0 +1,173 @@
+## Install
+
+```sh
+$ npm install --save jest-parallel
````
Just... no!

cpojer commented on packages/jest-parallel/README.md:
```diff
@@ -0,0 +1,173 @@
+Allow customizing all options passed to `childProcess.fork`. By default, some values are set (`cwd` and `env`), but you can override them and customize the rest. For a list of valid values, check [the Node documentation](https://nodejs.org/api/child_process.html#child_process_child_process_fork_modulepath_args_options).
+
+
+### `getWorkerKey: (method: string, ...args: Array<any>) => ?string` (optional)
```
I would go with @davidaurelio's suggestion and use `computeWorkerKey` because this function is potentially expensive, and it would be good to denote that in the name.

cpojer commented on packages/jest-parallel/README.md:
```diff
@@ -0,0 +1,173 @@
+  // Wait a bit.
+  await sleep(10000);
+
+  // Transform again the same file. Will immediately return because the
```
"Transform the same file again"

cpojer commented on packages/jest-parallel/src/__performance_tests__/.gitignore:
```diff
@@ -0,0 +1 @@
+node_modules
```
Can you add this to the top level gitignore file instead? Thanks!

cpojer commented on packages/jest-parallel/package.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "jest-parallel",
+  "version": "20.0.4",
```
update to latest jest version pls.

cpojer commented on packages/jest-parallel/src/__performance_tests__/workers/jest_parallel.js:
```diff
@@ -0,0 +1,7 @@
+const pi = require('./pi');
+
+module.exports.loadTest = function() {
+  pi();
```
I think if you want to test perf accurately, you need to return the value here back to the parent process.

cpojer commented on packages/jest-parallel/src/types.js:
```diff
@@ -0,0 +1,94 @@
+
+/* eslint-disable no-unclear-flowtypes */
+
+export type JestApi = {[string]: (...any) => Promise<any>};
```
Can you drop "Jest" from the types everywhere? I don't really think that adds much.

cpojer commented on packages/jest-parallel/src/index.js:
```diff
@@ -0,0 +1,182 @@
+    const aggregatedStderr = mergeStream();
+    const workers = [];
+
+    if (!path.isAbsolute(options.workerPath)) {
```
could we instead just run "require.resolve" here ourselves? It will throw if the module isn't found, and it allows people to pass `my-worker-package` which may be a form of a "plugin API" for workers some day.

mjesun replied:
We can't. `require.resolve` is relative to the file calling it. If we call it ourselves, all paths will always be relative to `jest-parallel` itself, which is undesirable.

cpojer replied:
Let's call require.resolve anyway as it'll work with packages (like 'my-worker-package') and throw otherwise?

SimenB replied:
Would https://github.com/sindresorhus/resolve-from help?

cpojer replied:
Nah, I think that's going too far. It should either be a pre-resolved module or a node_module.

mjesun replied:
`resolve-from` would work, but you need to go one level up the stack to check who called into `jest-parallel`. I really dislike that approach.

cpojer commented on packages/jest-parallel/src/index.js:
```diff
@@ -0,0 +1,182 @@
+    return this._api;
+  }
+
+  getAggregatedStdout(): Readable {
```
I think just `getStdout` is sufficient as a name for this, no? It's implied that it is the aggregated data.

cpojer commented on packages/jest-parallel/src/index.js:
```diff
@@ -0,0 +1,182 @@
+  _aggregagedStderr: Readable;
+  _api: JestApi;
+  _ending: boolean;
+  _hashes: {[string]: JestWorker};
```
nit: rename to "_keys" or "_cacheKeys"

cpojer commented on packages/jest-parallel/src/index.js:
```diff
@@ -0,0 +1,182 @@
+    };
+
+    for (let i = 0; i < options.workers; i++) {
+      const jestWorker = new JestWorker(workerOptions);
```
remove the "jest" everywhere.

cpojer commented on packages/jest-parallel/src/child.js:
```diff
@@ -0,0 +1,95 @@
+
+  process.send({
+    result,
+    type: 'ok',
```
instead of a string, could we use a number, like 0 for success and 1 for error? I want to make sure the overhead is as minimal as possible. It should probably be `[0, result]`.

cpojer commented on packages/jest-parallel/src/child.js:
```diff
@@ -0,0 +1,95 @@
+  if (result && typeof result.then === 'function') {
+    result.then(reportSuccess, reportError);
+  } else {
+    reportSuccess(result);
```
is there any possibility that `reportSuccess` could throw and we'd need to handle an error here?

cpojer commented:
This is really great @mjesun, and really solid work. I'm going to propose a number of API changes that I think make this module much nicer to work with, and sets it further apart from worker-farm (I don't like the "farm" word in general).

What do you think about this simple high-level API?

```javascript
const Worker = require('jest-parallel');
const worker = new Worker(path, {
  computeWorkerKey: () => {},
  …
});

await worker.transform(file);

worker.getStdout();
worker.end();
```

This get's rid of the word "Farm" and instead calls the overall class the "worker". This is similar to the web worker API in a way, but the idea is that `jest-parallel` hints at the fact that the module parallelizes work, and the interface we provide should be that of a worker, regardless of how many are actually working concurrently. I think that's smooth.

Then, second, let's get rid of `getAPI`. The new `Worker` class only provides a small number of methods: let's make it so that `end`, `getStdout`, `getStderr` and it will simply throw if the worker wants to define either of those. Those three aren't proper names for worker functions anyway, so there is little conflict potential.

I changed the API to demand the path as first argument, since it is always mandatory, and changed the cache key function to `computeWorkerKey` like described in my comment above.

Finally, and I'm not sure if we should go through with this, but it would be awesome if `jest-parallel` could infer the exposed methods on it's own without having to pass it. There are multiple ways to do it, with the most obvious one being to require the user's worker file in the constructor and using `Object.keys(require(workerPath))` and then plucking those on the individual worker instance. This wouldn't add any overhead if we recommend people to use inline-requires in worker files (I'd be totally fine with that). If we want to maintain the same behavior as before, we could be a bit more crazy and use Proxy to call through to the worker in the first `await worker.transform()` call, return the exposed methods from the child, and if it matches it would go through with the computation, pluck the right methods in and otherwise throw that `worker.transform` is not defined. Having the source of truth in only one place (the worker file itself) and without an `Array<string>` definition that needs to be maintained, this API would truly be super simple to use. I really think that would be awesome.

What do you think about this proposal? I think that significantly simplifies the API and makes it more approachable, which should encourage people to use `jest-parallel` more to make their tools faster, which is the goal here :)

orta commented:
I like the new API idea, nit-pick: `worker.transform`might make more sense as `worker.transact` as you want the naming to be use-case agnostic.

cpojer commented:
@orta keep in mind that those functions would be dynamic and come from the user-specified worker function. They should map directly to what that module exports. `exports.foo` would make `worker.foo` available.

mjesun commented:
@cpojer I'm in for all the described changes and I like them, except for the last one. I would love to drop the fact that `exposedMethods` needs to be passed, and I tried the `Proxy` approach, but it makes things slower. I guess the question here is whether we want to sacrifice performance by commodity; in my case I think it's not a big problem for the user to have to manually define the methods (plus, the `get` `Proxy` trap is generically not pollyfillable).

Regarding an auto-`require` on the main thread, it could cause a memory leak, or even a block of the main thread depending on what the worker is expected to do (technically, it can't block it, because it would not be able to receive anything on the child process if so; but it can render the process useless by stealing almost all cycles).

I will implement the rest of the suggestions while we agree on something for the auto-discovery.

cpojer commented:
Yeah, you are right, Proxy is probably overkill.

How about we do this: the default method is that we require the module in the parent context and look at what it exports, but we retain the option to be explicit about the callable API which means we'd skip the in-process detection?

Another path I see is that we could add static analysis to extract `exports` using the AST of the path module. The overhead of that should be minimal, and ES module exports are static. The downside is that it would only work with ES modules, but since that is where we are headed and this is an ultra modern approach :P That one of course also comes with retaining the option to explicitly specify the methods, I just like the idea of inferring the API methods to make this API super simple.

mjesun commented:
@cpojer @SimenB Most of the feedback (except `Proxy`) got added; since there's been a lot of changes I'd like you to review the PR again as if it was new 🙂.

* Worker path is now the first, only required argument of the constructor. All options are optional.
* All proposed renamings were accepted (remove `aggregated` from streams, `cacheKeys`, `computeWorkerKey`...).
* The `getApi` method is removed.
* Auto-discovery for methods, if `exposedMethods` is not provided as an option. This is done by requiring the file in the main process. Support for an exported method is also available.
* `Array`-ified all messages travelling back and forth between the main process and workers.
* Updated the `README` to include API changes, re-write part of it and add a minimal example section that demonstrates how to use it with two lines.
* Added tests, so coverage is still at 100%, including all lines, statements, and branches.

There is one last additional change that we will make later, which is renaming `jest-parallel` itself to `jest-worker`. That one will come in an additional PR.

mjesun commented:
More changes done:

* Renamed module from `jest-parallel` to `jest-worker`.
* Renamed `workers` to `numWorkers` to avoid confusion between the list of workers and its amount.
* Updated coverage image on the PR description and performance analysis, based on the latest changes. Looks like we got some performance improvements by switching to arrays + using integers for message signaling.

cpojer commented on packages/jest-worker/src/index.js:
```diff
@@ -0,0 +1,193 @@
+ *   processed by the same worker. This is specially useful if your workers are
+ *   caching results.
+ */
+class Farm {
```
Let's rename this to "Worker" like we discussed yesterday.

mjesun replied:
I thought we agreed on using `Worker` for the farm just externally.

cpojer commented on packages/jest-worker/src/index.js:
```diff
@@ -0,0 +1,193 @@
+      const child = require(workerPath);
+
+      exposedMethods = Object.keys(child)
+        .filter(name => this._isValidMethod(name))
```
This makes it silently ignore `end`. Should we throw if `end` is provided?

mjesun replied:
I thought about that; but what would happen if you require a module who exposes an `end` method, or any other underscored property we would crash. If you have a strong opinion on that I can remove the filter for validity.

cpojer replied:
I think we should throw, and in those cases demand people to use `exposedMethods` or wrap their module. Otherwise it's super confusing.

mjesun replied:
Oki!

cpojer commented on packages/jest-worker/src/index.js:
```diff
@@ -0,0 +1,193 @@
+
+  /* istanbul ignore next */
+  _empty() {
+    // Intentionally empty method.
```
what is this for and why is it here?

mjesun replied:
This is a nice way of creating an empty method only once.

cpojer replied:
`const emptyFn = () => {};` instead of the instance method.

cpojer commented:
Sick! I love the API. I'll merge it either when CI finishes or you make another iteration based on my small comments (otherwise please do follow-ups).

One thing we leverage in Jest is transient failures/retries. Is jest-worker going to implement this in a follow-up or are we going to push the responsibility up the stack?

SimenB commented on packages/jest-worker/README.md:
```diff
@@ -0,0 +1,193 @@
+
+List of method names that can be called on the child processes from the parent process. You cannot expose any method named like a public `Worker` method, or starting with `_`. If you use method auto-discovery, then these methods will not be exposed, even if they exist.
+
+#### `numWorkers: number` (required)
```
It says "required" in the heading, but "defaults" in the body. Which is it?

mjesun replied:
Good catch! It's optional.

SimenB commented on packages/jest-worker/src/__tests__/child.test.js:
```diff
@@ -0,0 +1,270 @@
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
```
MIT?

SimenB commented on packages/jest-worker/src/worker.js:
```diff
@@ -0,0 +1,163 @@
+
+'use strict';
+
+const childProcess = require('child_process');
```
`import`

SimenB commented on packages/jest-worker/src/worker.js:
```diff
@@ -0,0 +1,163 @@
+  }
+}
+
+module.exports = Worker;
```
`export default`

SimenB commented on packages/jest-worker/package.json:
```diff
@@ -0,0 +1,13 @@
+    "type": "git",
+    "url": "https://github.com/facebook/jest.git"
+  },
+  "license": "BSD-3-Clause",
```
MIT. All headers as well

cpojer commented:
Oh yeah, we switched from BSD+Patents to MIT license since your PTO.

mjesun commented:
> One thing we leverage in Jest is transient failures/retries. Is jest-worker going to implement this in a follow-up or are we going to push the responsibility up the stack?

When a worker fails (OOM, the process exits, throws...), a new worker is spawned to cover the empty gap left by that worker that died. Calls are not removed from the queue (and not marked as processed) up until we get a result back, so as soon as the new worker is up, the last call will be passed into it again.

cpojer commented:
What happens if the same call to the worker throws every time?

mjesun commented:
A call to the worker throwing does not mean the process exits; so nothing happens. If the process is consistently dying, it is re-spawned all the time; but that's a failure on the `jest-worker` level and not on the "forked module" level.

SimenB commented:
Sounds like it should only retry a set amount of times before throwing, so we avoid it retrying forever and making the process seemingly hang

cpojer commented:
Exactly, `worker-farm` throws after like 2 retries and exits, I think we should retain that.

mjesun commented:
I'm not very sure about that. If we have workers OOMing and we don't care about that, after three OOMs we would kill the worker without further notice, and you're trapped. Adding an option as well to allow the amount of times this can happen would complicate the API.

cpojer commented:
@mjesun I'm not sure I follow. With the code as you explained it, it'll cause an infinite loop of the child crashing and the parent retrying the same thing. Keep in mind we don't have control over the code that is run in the worker (for example with Jest we run user code in tests in child processes). Right now, if one test crashes the process, we retry twice (could be an intermittent failure) and then worker-farm bails and says it cannot retry. What will the behavior be with `jest-worker`?

mjesun commented:
What I mean is that when a child crashes is not because a method called threw; it is because something, either on the `jest-worker` side, or on the Node internals, made it crash. In that case, I think it's ok to re-spawn the thread. If the thread keeps dying, it's almost certainly because of a bug on `jest-worker` and not on the child module itself.

cpojer commented:
Let’s say you have a Jest Test that accidentally calls “process.exit()” or dies in some other way. With jest-worker, how would you handle such a test? Worker farm will retry and eventually crash Jest.

________________________________
From: Miguel Jiménez Esún <notifications@github.com>
Sent: Tuesday, October 3, 2017 5:50:05 PM
To: facebook/jest
Cc: Christoph Nakazawa; Mention
Subject: Re: [facebook/jest] Initial version of jest-worker (#4497)


What I mean is that when a child crashes is not because a method called threw; it is because something, either on the jest-worker side, or on the Node internals, made it crash. In that case, I think it's ok to re-spawn the thread. If the thread keeps dying, it's almost certainly because of a bug on jest-worker and not on the child module itself.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/facebook/jest/pull/4497#issuecomment-333907652>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAA0KEgWrmuUf33NFjN7KisfU_N36QtOks5somW9gaJpZM4PaMqp>.

cpojer commented on packages/jest-worker/src/child.js:
```diff
@@ -68,11 +68,16 @@ function reportError(error: Error) {
     PARENT_MESSAGE_ERROR,
-    error.constructor.name,
+    error.constructor && error.constructor.name,
     error.message,
```
What happens when a string is thrown? `throw 'Foo'`?

mjesun replied:
I just added a commit to allow this.

cpojer commented:
Let's ship it.

pedrottimark commented:
Awesome work! For your info, after `yarn clean-all && yarn` local `yarn jest` fails:

```sh
FAIL  packages/jest-worker/src/__tests__/child.test.js
 ● lazily requires the file

   TypeError: Cannot read property 'concat' of undefined

     at handle (node_modules/worker-farm/lib/child/index.js:44:24)
     at emitOne (events.js:120:20)
     at process.emit (events.js:210:7)
         at Promise (<anonymous>)
```

MacOS 10.12.6
yarn 1.1.0
node 8.2.0 and 6.11.3 and 4.8.4

However `yarn jest child` succeeds.

mjesun commented:
Interesting. It might be related to the fact that `child.js` messes with `process.send`, which is used also by the internal Jest worker (right now, `worker-farm`, soon `jest-worker`) to perform IPC communication.

I also wonder why CI environments, which are always starting from a clean environment, did not catch it. I will try to reproduce locally and find a fix if it reproduces. Thanks for noticing!

pedrottimark commented:
There is vague memory of earlier discussion about number of threads testing locally versus CI?
